Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,105 @@
+#if !NET40
+namespace Microsoft.ApplicationInsights.DependencyCollector.Implementation
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Globalization;
+    using System.Linq;
+    using System.Net;
+    using Microsoft.ApplicationInsights.Common;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Web.Implementation;
+
+    /// <summary>
+    /// Diagnostic listener implementation that listens for Http DiagnosticSource to see all outgoing HTTP dependency requests.
+    /// </summary>
+    internal class HttpDiagnosticSourceListener : IObserver<KeyValuePair<string, object>>, IDisposable
+    {
+        private readonly FrameworkHttpProcessing httpProcessingFramework;
+        private HttpDiagnosticSourceSubscriber subscribeHelper;
+        private bool disposed = false;
+
+        internal HttpDiagnosticSourceListener(FrameworkHttpProcessing httpProcessing)
+        {
+            this.httpProcessingFramework = httpProcessing;
+            this.subscribeHelper = new HttpDiagnosticSourceSubscriber(this);
+        }
+
+        /// <summary>
+        /// IDisposable implementation.
+        /// </summary>
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// This method gets called once for each event from the Http DiagnosticSource.
+        /// </summary>
+        /// <param name="value">The pair containing the event name, and an object representing the payload. The payload
+        /// is essentially a dynamic object that contain different properties depending on the event.</param>
+        public void OnNext(KeyValuePair<string, object> value)
+        {
+            try
+            {
+                dynamic propertyBag = value.Value;                        
+                switch (value.Key)
+                {
+                    case "System.Net.Http.Request":
+                        this.httpProcessingFramework.OnRequestSend((HttpWebRequest)propertyBag.Request);
+                        break;
+                    case "System.Net.Http.Response":
+                        this.httpProcessingFramework.OnResponseReceive((HttpWebRequest)propertyBag.Request, (HttpWebResponse)propertyBag.Response);
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                // This shouldn't happen, but just in case it does, we shouldn't break the application.
```

Comment:
as far as i can tell, all exceptions are handled in OnBegin/OnEnd.
However if there is a reason to catch them, consider writing error to the event log